hazelcast-jet-sql/src/main/java/com/hazelcast/jet/sql/impl/opt/logical/ProjectIntoScanLogicalRule.java

```diff
@@ -0,0 +1,279 @@
+/*
+ * Copyright (c) 2008-2020, Hazelcast, Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hazelcast.jet.sql.impl.opt.logical;
+
+import com.hazelcast.jet.sql.impl.opt.OptUtils;
+import com.hazelcast.sql.impl.calcite.schema.HazelcastTable;
+import org.apache.calcite.plan.RelOptRule;
+import org.apache.calcite.plan.RelOptRuleCall;
+import org.apache.calcite.rel.core.Project;
+import org.apache.calcite.rel.core.RelFactories;
+import org.apache.calcite.rel.core.TableScan;
+import org.apache.calcite.rel.logical.LogicalProject;
+import org.apache.calcite.rel.logical.LogicalTableScan;
+import org.apache.calcite.rex.RexCall;
+import org.apache.calcite.rex.RexInputRef;
+import org.apache.calcite.rex.RexNode;
+import org.apache.calcite.rex.RexShuttle;
+import org.apache.calcite.rex.RexVisitorImpl;
+import org.apache.calcite.util.mapping.Mapping;
+import org.apache.calcite.util.mapping.Mappings;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Map;
+
+/**
+ * Logical rule that pushes down column references from a {@link Project}
+ * into a {@link TableScan} to allow for constrained scans. See {@link HazelcastTable}
+ * for more information about constrained scans.
+ * <p>
+ * <b>Case 1: </b>projects that have only column expressions are eliminated completely,
+ * unused columns returned from the {@code TableScan} are trimmed.
+ * <p>
+ * Before:
+ * <pre>
+ * LogicalProject[projects=[$2, $1]]]
+ *   LogicalTableScan[table[projects=[0, 1, 2]]]
+ * </pre>
+ * After:
+ * <pre>
+ * LogicalTableScan[table[projects=[2, 1]]]
+ * </pre>
+ * <b>Case 2: </b>projects with non-column expressions trim the unused columns only.
+ * <p>
+ * Before:
+ * <pre>
+ * LogicalProject[projects=[+$2, $0]]]
+ *   LogicalTableScan[table[projects=[0, 1, 2]]]
+ * </pre>
+ * After:
+ * <pre>
+ * LogicalProject[projects=[+$0, $1]]]
+ *   LogicalTableScan[table[projects=[2, 0]]]
+ * </pre>
+ */
+final class ProjectIntoScanLogicalRule extends RelOptRule {
```

Comment:
From what I understand, you decided to implement this rule to trim columns only, similarly to IMDG part. Generally, many (if not all) logical rules are going to be exactly the same between IMDG and Jet. Have you considered using IMDG rules? If not, do you plan to merge the rule sets eventually?

Reply:
I compared the two files. Besides formatting there are these differences:
- Jet uses `operand(TableScan.class, any())`, imdg uses `operandJ(LogicalTableScan.class, null, OptUtils::isHazelcastTable, none())`. I think the additional predicate is ok for us.
- jet uses `OptUtils.createLogicalScan`, imdg uses `OptUtils.createLogicalScanWithNewTable`. I guess there's some issue fixed in the imdg variant

I tried to drop Jet's rule and use imdg's all tests passed. Did the same for `FilterIntoSchanLogicalRule`, the differences were the same. So I committed it.